modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/validation/OAuthConnectionProviderModelValidator.java

```diff
@@ -46,27 +53,62 @@ protected void onConnectionProvider(ConnectionProviderModel model) {
           return;
         }
 
-        AuthorizationCode authCode = implementingType.getAnnotation(AuthorizationCode.class);
-        ClientCredentials clientCredentials = implementingType.getAnnotation(ClientCredentials.class);
+        Optional<OAuthModelProperty> optionalOAuthModelProperty = model.getModelProperty(OAuthModelProperty.class);
 
-        if (authCode != null && clientCredentials != null) {
+        boolean supportsAuthCode = optionalOAuthModelProperty.map(oAuthModelProperty -> oAuthModelProperty.getGrantTypes()
+            .stream().filter(oAuthGrantType -> oAuthGrantType instanceof AuthorizationCodeGrantType).findFirst().isPresent())
+            .orElse(false);
+
+        boolean supportsClientCredentials = optionalOAuthModelProperty
+            .map(oAuthModelProperty -> oAuthModelProperty.getGrantTypes().stream()
+                .filter(oAuthGrantType -> oAuthGrantType instanceof ClientCredentialsGrantType).findFirst().isPresent())
+            .orElse(false);
+
+
+        if (supportsAuthCode && supportsClientCredentials) {
           throw new IllegalConnectionProviderModelDefinitionException(format(
                                                                              "Connection Provider of class '%s' is attempting to support both authorization code and client credentials "
                                                                                  + "grant types. Each connection provider can only support one grant type at a time.",
                                                                              implementingType));
         }
 
-        if (authCode != null) {
+        if (supportsAuthCode) {
           validateStateField(implementingType, AuthorizationCodeState.class, "authorization code");
         }
 
-        if (clientCredentials != null) {
+        if (supportsClientCredentials) {
           validateStateField(implementingType, ClientCredentialsState.class, "client credentials");
         }
+
```

Comment:
this block of code might also be moved to a separate function to reveal its intention and enhance readability, what do you think?

Reply:
agree